File: src/main/java/org/nia/logic/quests/judge/JudgeQuest.java
package org.nia.logic.quests.judge;

import org.nia.logic.quests.IQuest;
import org.nia.logic.quests.IQuestEvent;
import org.nia.logic.quests.IQuestStep;
import org.nia.logic.quests.QuestsEnum;

import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * @author Иван, 28.03.2017.
 */
public class JudgeQuest implements IQuest {
    public static JudgeQuest INSTANCE = new JudgeQuest();

    @Override
    public String getStart() {
        return "Остап попросил тебя заменить заболевшего городского юриста на сегодняшних заседаниях суда. " +
                "Ты берешь подобающий по случаю парик и идешь в здание суда. В зависимости от желаний судьи и " +
                "потерпевшего, тебе придется исполнять роль прокурора или адвоката. Награда за " +
                "задание будет увеличиваться или уменьшаться в зависимости от того, насколько успешно ты будешь вести дела.";
    }

    public enum JudgeEvent implements IQuestEvent {
        JUDGE_FIELD_ROWS_JUDGEMENT(FieldRowsJudgement.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return FieldRowsJudgement.valueOf(questStep);
            }

            @Override
            public boolean canBeRandomed() {
                return false;
            }
        },
        JUDGE_STOLEN_HORSE(StolenHorse.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return StolenHorse.valueOf(questStep);
            }
        },
        JUDGE_BROTHERS(Brothers.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return Brothers.valueOf(questStep);
            }
        }
        ;
        private IQuestStep init;

        JudgeEvent(IQuestStep init) {
            this.init = init;
        }

        @Override
        public IQuestStep getInit() {
            return init;
        }

        @Override
        public String getName() {
            return name();
        }

        @Override
        public QuestsEnum getQuestsEnum() {
            return QuestsEnum.JUDGE;
        }
    }

    @Override
    public IQuestEvent getRandomEvent() {
        List<JudgeEvent> events = Arrays.stream(JudgeEvent.values())
                .filter(IQuestEvent::canBeRandomed)
                .collect(Collectors.toList());
        return events.get(new Random().nextInt(events.size()));
    }

    @Override
    public IQuestEvent getEvent(String event) {
        return JudgeEvent.valueOf(event);
    }
}

File: src/main/java/ru/nia/tavern/quests/judge/JudgeQuest.kt
package ru.nia.tavern.quests.judge

import ru.nia.tavern.quests.IQuest
import ru.nia.tavern.quests.IQuestEvent
import ru.nia.tavern.quests.IQuestStep
import ru.nia.tavern.quests.QuestsEnum
import java.util.*
import java.util.stream.Collectors

/**
 * @author Иван, 28.03.2017.
 */
class JudgeQuest : IQuest {

    override val start: String
        get() = "Остап попросил тебя заменить заболевшего городского юриста на сегодняшних заседаниях суда. " +
                "Ты берешь подобающий по случаю парик и идешь в здание суда. В зависимости от желаний судьи и " +
                "потерпевшего, тебе придется исполнять роль прокурора или адвоката. Награда за " +
                "задание будет увеличиваться или уменьшаться в зависимости от того, насколько успешно ты будешь вести дела."

    override val randomEvent: IQuestEvent
        get() {
            val events = Arrays.stream(JudgeEvent.values())
                    .filter(Predicate<JudgeEvent> { it.canBeRandomed() })
                    .collect<List<JudgeEvent>, Any>(Collectors.toList())
            return events[Random().nextInt(events.size)]
        }

    enum class JudgeEvent private constructor(override val init: IQuestStep) : IQuestEvent {
        JUDGE_FIELD_ROWS_JUDGEMENT(FieldRowsJudgement.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return FieldRowsJudgement.valueOf(questStep)
            }

            override fun canBeRandomed(): Boolean {
                return false
            }
        },
        JUDGE_STOLEN_HORSE(StolenHorse.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return StolenHorse.valueOf(questStep)
            }
        },
        JUDGE_BROTHERS(Brothers.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return Brothers.valueOf(questStep)
            }
        };

        override val name: String
            get() = name

        override val questsEnum: QuestsEnum
            get() = QuestsEnum.JUDGE
    }

    override fun getEvent(event: String): IQuestEvent {
        return JudgeEvent.valueOf(event)
    }

    companion object {
        var INSTANCE = JudgeQuest()
    }
}
